### PR TITLE
real cost for Nevlas with PI

### PR DESCRIPTION
**Pull request by rzulian**

### CODE REVIEW

File: src/reward.ts [hunk 1]
```diff
@@ -69,6 +69,11 @@ export default class Reward {
     return rewards.map(reward => new Reward(-reward.count, reward.type));
   }
 
+  static discount(rewards: Reward[], multiplier: number, type: Resource): Reward[] {
```
Comment (coyotte508) [inline src/reward.ts]: Should add a comment, like:

`/** For each reward, if it is of corresponding type, reduce the cost by dividing it by multiplier */`

I would also switch the order of the type / multiplier arguments.

Comment (rzulian) [reply src/reward.ts]: @coyotte508 : are you using ```tokenModifier``` in the expansion? If not, I would remove it as it's used only for Nevlas with PI

Comment (coyotte508) [reply src/reward.ts]: Nope, not using it. For now it could be replaced with faction === Nevlas ? 2 : 1.

Later we can make it more flexible if need be.

Comment (coyotte508) [reply src/reward.ts]: I need this PR for the range extension when moving spaceships in the expansion (it uses power)

**Comment by rzulian** [comment]
@coyotte508 : I've removed ```tokenModifier``` and added a ```discounted``` property in the boardActions. I've left the ```discount``` method in the ```reward```.
The real problem I see is that this is not backward compatible.

**Comment by coyotte508** [comment]
Yea. Also it means that everywhere that power can be spent, code needs to be added specifically for nevlas (e.g. new board actions, ...) as `spendablePowerTokens()` doesn't take into account the discount. This is a small technical debt for the future. Taklons don't have that issue. 

I think the main issue was the visualisation, and that's why the issue was opened in gaia-viewer at first. Maybe the new player info design will make it a bit less confusing, since there's no 'pw' text in it but three drawn bowls? 

Actually, only now I think that a much simpler solution would be to change ` 0/2/2 pw` to something like` 0/2/2(=4) pw` in the viewer for nevlas.

I'm really sorry about realizing this now and dragging this out until the PR is ready.

**Comment by rzulian** [comment]
The only missing visualisation in viewer is actually only the hint in the boardActions, which has to be different when Nevlas with PI.
Managing th legacy games is a different issue. we can either do conditional code, or repairs old games (which I had to do for tests).
On the conditional code we could ad a ```EngineVersion``` property on the game. What do you think?

File: src/actions.ts [hunk 2]
```diff
@@ -34,23 +34,32 @@ export const freeActionsNevlas = [
...
-  [BoardAction.Qic2]: { cost: "3q", income: [">fed"] },
-  [BoardAction.Qic3]: { cost: "2q", income: ["3vp", "pt > vp"]}
+  [BoardAction.Power1]: { cost: "7pw", discounted: "4pw", income: ["3k"] },
+  [BoardAction.Power2]: { cost: "5pw", discounted: "3pw", income: ["2step"] },
+  [BoardAction.Power3]: { cost: "4pw", discounted: "2pw", income: ["2o"] },
+  [BoardAction.Power4]: { cost: "4pw", discounted: "2pw", income: ["7c"] },
+  [BoardAction.Power5]: { cost: "4pw", discounted: "2pw", income: ["2k"] },
+  [BoardAction.Power6]: { cost: "3pw", discounted: "2pw", income: ["1step"] },
+  [BoardAction.Power7]: { cost: "3pw", discounted: "2pw", income: ["2t"] },
+  [BoardAction.Qic1]: { cost: "4q", discounted: "4q", income: ["tech"] },
```
Comment (coyotte508) [inline src/actions.ts]: What happens if there's a faction that discounts qic usage (eg -1qic) for board actions?

File: src/actions.ts [hunk 3]
```diff
@@ -34,23 +34,32 @@ export const freeActionsNevlas = [
...
-  [BoardAction.Power2]: { cost: "5pw", income: ["2step"] },
-  [BoardAction.Power3]: { cost: "4pw", income: ["2o"] },
-  [BoardAction.Power4]: { cost: "4pw", income: ["7c"] },
-  [BoardAction.Power5]: { cost: "4pw", income: ["2k"] },
-  [BoardAction.Power6]: { cost: "3pw", income: ["1step"] },
-  [BoardAction.Power7]: { cost: "3pw", income: ["2t"] },
-  [BoardAction.Qic1]: { cost: "4q", income: ["tech"] },
-  [BoardAction.Qic2]: { cost: "3q", income: [">fed"] },
-  [BoardAction.Qic3]: { cost: "2q", income: ["3vp", "pt > vp"]}
+  [BoardAction.Power1]: { cost: "7pw", discounted: "4pw", income: ["3k"] },
```
Comment (coyotte508) [inline src/actions.ts]: I said that nevlas were the only faction to discount power even in the expansion, but that was assuming a call to something like `discount(rewards, Resource.Power, nevlasDiscount ? 2 : 1)` that could be made more flexible in the future. Not hardcoding the discounted price with a fixed multiplier.

File: src/faction-boards/nevlas.ts [hunk 4]
```diff
@@ -15,10 +15,10 @@ const nevlas: FactionBoardRaw = {
   },
   income: ["2k,4o,15c,q,up-sci", "+o,k"],
   handlers: {
-    [`build-${Building.PlanetaryInstitute}`]: (player: Player) => player.data.tokenModifier = 2,
     freeActionChoice:  (player: Player, pool: any[]) => {
       pool.push(...freeActionsNevlas);
       if (player.data.hasPlanetaryInstitute()) {
+        pool.length = 0;
```
Comment (coyotte508) [inline src/faction-boards/nevlas.ts]: Nice. I didn't know arrays could be resized this way!

File: src/reward.ts [hunk 5]
```diff
@@ -69,6 +69,12 @@ export default class Reward {
     return rewards.map(reward => new Reward(-reward.count, reward.type));
   }
 
+  /** For each reward, if it is of corresponding type, reduce the cost by dividing it by multiplier */
+  static discount(rewards: Reward[], type: Resource, multiplier: number): Reward[] {
```
Comment (coyotte508) [inline src/reward.ts]: Sadly it seems this function is now unused

File: src/actions.ts (hunk 3 above)
Comment (rzulian) [reply src/actions.ts]: Yes, probably your suggestion is better

File: src/faction-boards/nevlas.ts (hunk 4 above)
Comment (rzulian) [reply src/faction-boards/nevlas.ts]: Actually, that was the only way to let it work. Everything else wasn't resetting ```pool``` in the emit event :)

File: src/reward.ts (hunk 5 above)
Comment (rzulian) [reply src/reward.ts]: Unless we use your suggestion above

File: src/faction-boards/nevlas.ts (hunk 4 above)
Comment (coyotte508) [reply src/faction-boards/nevlas.ts]: Probably `pool.splice(0)` would have worked. But your way is nicer :)

**Comment by rzulian** [comment]
We can put this on hold until we decide what to do. 
The problem I see is that our logs are wrong. We will see.
screen shot 2018-10-14 at 18 43 41

**Pull request closed without merging**